### DISCUSSION TITLE
Getting started

### DISCUSSION THREAD

**Discussion by bigcurl** [bigcurl/ffi-cuda-toolkit-gem-generator#2] [General]
Hi @s96Abrar,

If you still intend to join working on this project here are some links to get you started in the Ruby world and clarify some terminology.

**Ruby**
- https://learnxinyminutes.com/docs/ruby/
- https://www.ruby-lang.org/en/documentation/ruby-from-other-languages/to-ruby-from-c-and-cpp/
- https://docs.ruby-lang.org/en/3.0.0/
- https://www.codecademy.com/learn/learn-ruby/modules/learn-ruby-introduction-to-ruby-u/cheatsheet
- 

**Gem**
- https://guides.rubygems.org/make-your-own-gem/
- https://bundler.io/guides/creating_gem.html
- https://lokalise.com/blog/create-a-ruby-gem-basics/
- https://blog.engineyard.com/how-to-build-your-own-gem-in-ruby

**FFI**
- https://en.wikipedia.org/wiki/Foreign_function_interface
- https://github.com/ffi/ffi
- https://github.com/ffi/ffi/wiki
- https://www.rubyguides.com/2019/05/ruby-ffi/
- https://www.endpoint.com/blog/2018/04/ffi-in-ruby/

**Bundler**
- https://bundler.io/v2.2/#getting-started

**CUDA**
- https://docs.nvidia.com/cuda/
- https://docs.nvidia.com/cuda/archive/

Basic ideas of the existing code:
First I downloaded the documentation and put it into /vendor
From there I parse the docu and generate ruby code to interface with the FFI gem
Run the code by calling ruby bin/g.rb generate driver-api. This puts the code into /gem/cuda
From there the the CUDA call should be callable.

I will write more tomorrow about the goals, etc.

**Comment by s96Abrar** [comment]
OK. I will dive into the references you mentioned.

**Reply by s96Abrar** [reply]
After digging into the resources i found out that creating a gem is pretty easy and loading c library with ffi is also easy (Most of them already done by you).

And seems like you are already added most of the library interfaces in ruby. 
Where can i start from now? What is left or needed to be tested?

Ah everything is automated.

**Reply by bigcurl** [reply]
> After digging into the resources i found out that creating a gem is pretty easy and loading c library with ffi is also easy (Most of them already done by you).

Dont worry there is still enough todo :-)
 
> And seems like you are already added most of the library interfaces in ruby. Where can i start from now? What is left or needed to be tested?

I started with the driver api from Cuda to get a first version into a workable state and validate the idea of using FFI. 
There are many more apis in the documentation. Most important the runtime api and the cuFile API. But lets start small

So what is left todo in the driver api. You mentioned an interesting topic, versioning. It would be great to support more than one version of CUDA. I'm not sure for now what is the best way to do it in the gem. A gem release per version, a single gem for all versions which can import different versions during runtime? Suggestions are welcome.

For now the download of the documentation has to be done manualy. You find the archive here:

- https://docs.nvidia.com/cuda/archive/

I just dowloaded that into the repository and deleted all besides driver and runtime api.

Every method from the CUDA api should have at least one example in ruby code on how to use it. This demonstrates how to convert the values from ruby memory model to the C memory mode. Best place for this are the tests I guess. For testing I setup the minitest gem:

- https://semaphoreci.com/community/tutorials/getting-started-with-minitest
- https://github.com/seattlerb/minitest
- https://minitest.rubystyle.guide/

I will update one example and you can try if you get it run and continue from there.

> Ah everything is automated.

That is the main idea generate a gem by parsing the docu and let the parser generate the FFI code on the fly. If a new versions comes out only the outliers which can not be parsed need to be fixed.
I will mark the current outliers in comments by version. For example some return values are not correct in the docu in 11.2 but maybe NVIDIA will fix it in a newer version.

So what is left for now. I think the functions part of the driver api is almost complete, It is not using the enum types yet and just marks everything as a pointer in FFI. See here for more about that:

https://github.com/ffi/ffi/wiki/Enums

Missing are parsing the structs, the defines, the typedefs and writings the tests/examples. Than of course the versioning of the gem is open. I would not mind having a single gem for all versions.

Why don't you start with parsing the structs. Some code is already there but I'm not sure if it is sufficient as the header file for the structs (can be found in the cuda release /include folder) mentions that sometimes a struct is a mixture of values, structs and unions. But my C knowledge is not good enough to know the exact difference and how it is represented in memory.  

I'm currently working on the enums.

Would that work for you?

**Reply by bigcurl** [reply]
Also I just checked I have CUDA 11.2 and CUDA 11.4 installed.The ruby code only parses CUDA 11.4.2 for now. So you can go ahead and we add versioning later. Easier to have one version running and than make it multiversion.

**Reply by s96Abrar** [reply]
> A gem release per version, a single gem for all versions which can import different versions during runtime? Suggestions are welcome.

It's possible to get a specific version documentation so downloading and extracting the documentation and generating the interface for ruby all can be done automated, some using shell scripts and some with ruby.

Well I tried to locate the library file path for now by this in the template...

```
binary_list = `whereis -b libcuda`.split(':')
binary_path = '/usr/lib/x86_64-linux-gnu/libcuda.so'
binary_path = binary_list[1].strip if binary_list.count > 1 && !binary_list[1].strip.empty?
ffi_lib binary_path
```

Also just to be clear, we are loading the cuda library (`libcuda.so`) not the cuda runtime toolkit (`libcudart.so`)?

**Reply by s96Abrar** [reply]
> I'm currently working on the enums.
> Would that work for you?

Yes it's fine.

I will try to see what can be done with define and typedefs.

`#define` could be simple as for function definition it would be a ruby function and for some constant it would be variable I guess.

And `typedefs` shown in [here](https://github.com/ffi/ffi/wiki/Examples#using-typedef)

**Reply by bigcurl** [reply]
Yep we are loading libcuda.so for the driver api and libcudart.so for the runtime api which comes later.

**Reply by bigcurl** [reply]
Here is a little refactoring from the loading code. Does the same thing than yours. Just strips first and this way uses a little easier to read condition. But like I said it should be easy for you writing ruby code.

```
binary_list = `whereis -b libcuda`.strip.split('libcuda: ')
binary_path = '/usr/lib/x86_64-linux-gnu/libcuda.so'
binary_path = binary_list[1] unless binary_list[1].nil?
ffi_lib binary_path
```

**Reply by s96Abrar** [reply]
Hmm... looks good.

**Comment by bigcurl** [comment]
Are you working on defines or typedefs? I would try the other than.

**Reply by s96Abrar** [reply]
Yes I will. I couldn't start yet.

I will try to do struct also.

**Reply by bigcurl** [reply]
Ah great. I have to leave for the day. I will see tomorrow where I can continue.
